IO tests, benchmark, refactor, and fixes

kprotty opened the pull request:
Fleshes out the state of io.zig, especially regarding networking

- [x] Move `io` stuff into its own directory
- [x] Add a customizable benchmark for socket interactions with io (there exists some already for fs interactions)
- [x] Move io tests into their own file + adds more integration tests
- [x] Fix IO.tick() on darwin

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,240 @@
+        transferred: u64 = 0,
+
+        // 1 MB: larger than socket buffer so forces io_pending on darwin
+        // Cnofigure this value to smaller amounts to test IO scheduling overhead
```
```suggestion
        // Configure this value to smaller amounts to test IO scheduling overhead
```

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,240 @@
+            const allocator = &gpa.allocator;
+            defer assert(gpa.deinit());
+
+            const buf = try allocator.alloc(u8, buffer_size * 2);
```
```suggestion
            const buffer = try allocator.alloc(u8, buffer_size * 2);
```

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,240 @@
+                .timer = &timer,
+                .started = started,
+                .current = started,
+                .tx = .{ .buffer = buf[0*buffer_size..][0..buffer_size] },
```
I think if we run `zig fmt` it will spread out the `0*buffer_size`? If not already run, let's quickly run `zig fmt` on this.

kprotty replied:
Yup, it did.

jorangreef commented on src/io/darwin.zig:
```diff
@@ -27,9 +29,7 @@ pub const IO = struct {
-        // TODO This is a hack to block 1ms every 10ms tick on macOS while we fix `flush(false)`:
-        // What we're seeing for the tigerbeetle-node client is that recv() syscalls never complete.
-        return self.run_for_ns(std.time.ns_per_ms);
+        return self.flush(false);
```
Good that this is done.

jorangreef commented on src/message_bus.zig:
```diff
@@ -164,14 +164,9 @@ fn MessageBusImpl(comptime process_type: ProcessType) type {
             // Set tcp recv buffer size
             if (config.tcp_rcvbuf > 0) rcvbuf: {
-                if (!is_darwin) {
+                if (is_linux) {
```
Much better now, stating the condition in the positive.

jorangreef reviewed (commented):
Thanks @kprotty. Are the only changes to `IO_Linux` the removal of `flags` and `openat`? Or were there any other changes that are not visible through the rename?

What do you think about defining `TCP_NODELAY` in IO or else putting it in `os.zig` so that `MessageBus` need only test if it's defined?

kprotty commented on src/io.zig:
```diff
@@ -1,994 +1,18 @@
-const IO_Uring = linux.IO_Uring;
-const io_uring_cqe = linux.io_uring_cqe;
-const io_uring_sqe = linux.io_uring_sqe;
+const testing = std.testing;
```
We don't. I'll be removing it

kprotty commented on src/io.zig:
```diff
@@ -1004,350 +28,6 @@ pub fn buffer_limit(buffer_len: usize) usize {
 }
 
 test "ref all decls" {
-    std.testing.refAllDecls(IO);
```
Don't think so as the tests are moved to their own file, not reached by the IO structs. Pushing up an alternative

kprotty commented:
> Are the only changes to `IO_Linux` the removal of `flags` and `openat`? Or were there any other changes that are not visible through the rename?

* `flags` were removed from the operations
* `openat` operation was removed
* `SOCK_CLOEXEC` flag default for `prep_accept`
* `MSG_NOSIGNAL` flag default for `prep_send/recv`
* `0` flag default for `prep_fsync`

> What do you think about defining `TCP_NODELAY` in IO or else putting it in `os.zig` so that `MessageBus` need only test if it's defined?

Since we use an older zig version, I'm not sure we can put it in the stdlib if this is what `os.zig` means. I'm for putting the definition in `IO` (similar to `openSocket`) if we may use the flag elsewhere. Decided to follow the flow of `Storage` and put it directly in `MessageBus` as a one-time-use thing for consistency. Thoughts?

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -125,7 +125,7 @@ pub fn main() !void {
         fn isRunning(self: Context) bool {
             // Make sure that we're connected
-            if (self.rx.socket.fd == -1) 
+            if (self.rx.socket.fd == -1)
```
For a branch like this, we can put this single-line statement also on the same line as the condition, without the newline.

Otherwise, if we need to span lines then we would use a block always.

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -166,12 +166,12 @@ pub fn main() !void {
             bytes: usize,
         ) void {
-            // The type of IO to perform 
+            // The type of IO to perform
```
Let's do this comment as a one-liner!

jorangreef commented on src/io/darwin.zig:
```diff
@@ -306,20 +306,20 @@ pub const IO = struct {
                     errdefer os.close(fd);
 
-                    // darwin doesn't support os.MSG_NOSIGNAL, 
+                    // darwin doesn't support os.MSG_NOSIGNAL,
```
Same here, this can be a one-liner.

jorangreef commented on src/io.zig:
```diff
@@ -27,7 +26,6 @@ pub fn buffer_limit(buffer_len: usize) usize {
-test "ref all decls" {
-    
+test "I/O" {
+    _ = @import("io/test.zig");
```
Cool

jorangreef reviewed (approved):
Thanks for the summary of changes. I made a few small suggestions.

Regarding `TCP_NODELAY`, first-prize long term is to suggest it for std lib `os.zig`, since it seems odd that we use things like `os.SO_RCVBUFFORCE` and `os.TCP_KEEPINTVL` but don't find `TCP_NODELAY` in there. Seems like a good candidate, unless I'm missing something. We can leave it in MessageBus as a stop-gap but would be great to be able to get rid of it when we update to 0.9.

kprotty commented:
> Seems like a good candidate, unless I'm missing something.

If `TCP_NODELAY` is recommended for stdlib, it would be available in 0.9.0/current-master release earliest. Tigerbeetle uses/downloads zig 0.8.0 (?) which makes me think it wouldn't get the stdlib definition if it was added.

jorangreef commented:
> If TCP_NODELAY is recommended for stdlib, it would be available in 0.9.0/current-master release earliest. Tigerbeetle uses/downloads zig 0.8.0 (?) which makes me think it wouldn't get the stdlib definition if it was added.

That's 100%. We'll update as soon as 0.9.0 is released (@ifreund has a 0.9 branch ready). We can stub it in the meantime, and then remove the stub once we update.

ifreund commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,238 @@
+const os = std.os;
+const assert = std.debug.assert;
+
+const Time = @import("time.zig").Time;
```
I think this needs to be `../time.zig`, same with the IO import below.

ifreund commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,238 @@
+            }
+
+            // Assert that everything is connected
+            try assert(self.server.fd != -1);
+            try assert(self.tx.socket.fd != -1);
+            try assert(self.rx.socket.fd != -1);
```
```suggestion
            assert(self.server.fd != -1);
            assert(self.tx.socket.fd != -1);
            assert(self.rx.socket.fd != -1);
```

ifreund commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,238 @@
+                        result: transfer_info.IoError!usize,
+                    ) void {
+                        const _bytes = result catch |err| {
+                            std.debug.panic("{} error: {}", .{ transfer_info.io_func, err });
```
Leaving off the specifier for slices is a compile error.
```suggestion
                            std.debug.panic("{s} error: {}", .{ transfer_info.io_func, err });
```

ifreund commented on src/io/benchmark.zig:
```diff
@@ -0,0 +1,238 @@
+const Time = @import("time.zig").Time;
+const IO = @import("io.zig").IO;
+
+pub fn main() !void {
+    try struct {
```
This `struct { fn run() void { ... } }.run();` pattern is nice for tests as it avoids polluting the outer scope. However I don't think it makes much sense for this benchmark code. Le'ts just do the standard
```zig
const Context = struct {};

pub fn main() void {
    // contents of the current run() function.
}
```

ifreund commented on src/io/test.zig:
```diff
@@ -0,0 +1,601 @@
+        fn run_test() !void {
+            const path = "test_io_write_fsync_read";
+            const file = try std.fs.cwd().createFile(path, .{ .read = true, .truncate = true });
+            // defer file.close(); // closed at the end
```
Let's not include commented code in the commits.

ifreund reviewed (changes requested):
The effect of the IO interface changes on the MessageBus code is a nice improvement :)

I found a few issues with the benchmark code, and there also seems to be a memory leak somewhere (thanks std gpa!):
```
/bin/zig run src/io/benchmark.zig --main-pkg-path src

IO throughput test: took 1000ms @ 10643.46 MB/s
thread 7722 panic: reached unreachable code
/usr/lib/zig/std/debug.zig:226:14: 0x21569b in std.debug.assert (benchmark)
    if (!ok) unreachable; // assertion failure
             ^
/home/ifreund/projects/tigerbeetle/src/io/benchmark.zig:40:25: 0x24403b in struct:9:9.run (benchmark)
            defer assert(gpa.deinit());
                        ^
/home/ifreund/projects/tigerbeetle/src/io/benchmark.zig:237:10: 0x23b4f0 in main (benchmark)
    }.run();
         ^
/usr/lib/zig/std/start.zig:471:37: 0x2337aa in std.start.callMain (benchmark)
            const result = root.main() catch |err| {
                                    ^
/usr/lib/zig/std/start.zig:413:12: 0x21760e in std.start.callMainWithArgs (benchmark)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/usr/lib/zig/std/start.zig:332:17: 0x216636 in std.start.posixCallMainAndExit (benchmark)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/usr/lib/zig/std/start.zig:245:5: 0x216442 in std.start._start (benchmark)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
zsh: abort      /bin/zig run src/io/benchmark.zig --main-pkg-path src
```

The more extensive and unified IO test coverage looks great!

kprotty commented:
Apparently `gpa.deinit()` returns true if there *was* a leak (no leak = correct = returns false) and assert checks for false...

The perils of lazily porting over test(ing.allocator) code to an executable.

kprotty commented:
I ran the benchmark locally on my system which has unusual overheads with io_uring in tigerbeetle (important context). I also switched the benchmark to use `run_for_ns(run_duration)` instead of a prng interleaving of `run_for_ns/tick` (which was primarily for testing as is already handled in `io/test.zig`). Finally, I removed the timer check in do_transfer() since run_for_ns now takes over the job of interrupting IO.

I found that the time to complete the benchmark became very variable: when `run_duration` was 1s, it ranged from 1s to upwards of 6s. I assume this was due to the amount of syscalls performed by the current `etime` machinery in flush(). I replaced this code with a timeout solution similar to `IO_Darwin` and the benchmark duration stabilized to run_duration.

The `etime` stuff appears to be there to handle completion/timeout races. I believe the darwin solution also works. I put these changes in a separate branch ([`io-cleanup-benchmark`](https://github.com/kprotty/tigerbeetle/tree/io-cleanup-benchmark)) in case we want to come back to this later. For now however, I'll just address the current review changes.

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -2,237 +2,237 @@ const std = @import("std");
+        transferred: usize = 0,
+    };
+
+    fn isRunning(self: Context) bool {
```
Let's do tiger_style: `is_running()`

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -2,237 +2,237 @@ const std = @import("std");
+    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
+    const allocator = &gpa.allocator;
+    defer {
+        const leaks = gpa.deinit();
+        assert(!leaks);
```
Nice and clear. It feels like `gpa.deinit()` should return false if it couldn't deinit completely (because of a leak).

jorangreef commented on src/io/benchmark.zig:
```diff
@@ -2,237 +2,237 @@ const std = @import("std");
+        const elapsed_ns = self.current - started;
+        const transferred_mb = @intToFloat(f64, self.transferred) / 1024 / 1024;
+
+        std.debug.warn("IO throughput test: took {}ms @ {d:.2} MB/s\n", .{
```
We should use `std.debug.print` here (or a stdout writer) instead since `warn` is deprecated?

jorangreef reviewed (approved):
🚀

jorangreef commented:
> I found that the time to complete the benchmark became very variable: when run_duration was 1s, it ranged from 1s to upwards of 6s. I assume this was due to the amount of syscalls performed by the current etime machinery in flush(). I replaced this code with a timeout solution similar to IO_Darwin and the benchmark duration stabilized to run_duration.
>
> The etime stuff appears to be there to handle completion/timeout races. I believe the darwin solution also works. I put these changes in a separate branch (io-cleanup-benchmark) in case we want to come back to this later. For now however, I'll just address the current review changes.

That's interesting. It would be good to understand exactly why the two approaches differ so much in variance. Thanks @kprotty for stashing this in a separate branch for reference.

I propose we come up with a minimal failing test case to reproduce the original I/O race condition at 1ns tick speeds, so we can confirm we have solved it. Once we understand it, we should then get together quickly with Jens to see what he thinks — he would appreciate the feedback, and would be able to flesh things out for us, especially with a simple test case.

The `run_for_ns` test case would also be really good to have even to prevent regressions.
